lib/Routing/Router.php

```diff
@@ -0,0 +1,116 @@
+
+use Doctrine\Website\Site;
+use Symfony\Component\Routing\Exception\ResourceNotFoundException;
+use Symfony\Component\Routing\Generator\UrlGenerator;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Routing\Matcher\UrlMatcher;
+use Symfony\Component\Routing\RequestContext;
+use Symfony\Component\Routing\Route;
+use Symfony\Component\Routing\RouteCollection;
+use function explode;
+use function parse_url;
+
+class Router
+{
+    /** @var RouteCollection */
+    private $routes;
+
+    /** @var RequestContext */
+    private $context;
+
+    /** @var UrlMatcher */
+    private $urlMatcher;
+
+    /** @var UrlGenerator */
+    private $urlGenerator;
+
+    /**
+     * @param mixed[] $routes
+     */
+    public function __construct(array $routes, Site $site)
+    {
+        $this->routes = new RouteCollection();
+
+        $this->context = $this->createRequestContext($site);
+
+        foreach ($routes as $routeName => $route) {
+            if (isset($route['controller']) && ! isset($route['defaults']['_controller'])) {
+                $route['defaults']['_controller'] = explode('::', $route['controller']);
+            }
+
+            if (isset($route['provider']) && ! isset($route['defaults']['_provider'])) {
+                $route['defaults']['_provider'] = explode('::', $route['provider']);
+            }
+
+            $route = new Route(
+                $route['path'],
+                $route['defaults'] ?? [],
+                $route['requirements'] ?? []
+            );
+
+            $this->routes->add($routeName, $route);
+        }
+
+        $this->urlMatcher   = new UrlMatcher($this->routes, $this->context);
+        $this->urlGenerator = new UrlGenerator($this->routes, $this->context);
+    }
+
+    public function getRouteCollection() : RouteCollection
+    {
+        return $this->routes;
+    }
+
+    /**
+     * @return mixed[]
+     */
+    public function match(string $pathinfo) : ?array
+    {
+        try {
+            return $this->urlMatcher->match($pathinfo);
+        } catch (ResourceNotFoundException $e) {
+            return null;
+        }
+    }
+
+    /**
+     * @param mixed[] $parameters
+     */
+    public function generate(
+        string $name,
+        array $parameters = [],
+        int $referenceType = UrlGeneratorInterface::ABSOLUTE_PATH
+    ) : string {
+        return $this->urlGenerator->generate($name, $parameters, $referenceType);
+    }
+
+    public function setContext(RequestContext $context) : void
+    {
+        $this->context = $context;
+    }
+
+    public function getContext() : ?RequestContext
+    {
+        return $this->context;
+    }
+
+    private function createRequestContext(Site $site) : RequestContext
+    {
+        $url = parse_url($site->getUrl());
+
+        return new RequestContext(
+            '',
+            'GET',
+            $url['host'],
+            $url['scheme'],
+            $url['port'],
```

Comment:
Can you make the port optional? This new version of the website doesn't allow `doctrine.website.url` without a port.